Update kernel and app classloader to be more efficient

jhanders34 opened the pull request:
- Update to not create locks for classes that will not be found by the boot classloader.  This reduces the amount of garbage created and reduces the amount of work that GC does since we use weak references to the classloader locks.
- Update loadClass to only lock around loading of the class from this loader.  In ParentLast load don't hold the classloader lock when calling the parent.  Also when calling the class generator don't hold the classloader lock.  It is not needed.
- Update to use replace(char, char) instead of replaceAll(String, String)

jhanders34 committed:
Update boot jar classloader to be more efficient

- Update to not create locks for classes that will not be found by the
boot classloader.  This reduces the amount of garbage created and
reduces the amount of work that GC does since we use weak references to
the classloader locks

dazavala commented on dev/com.ibm.ws.kernel.boot.core/src/com/ibm/ws/kernel/internal/classloader/JarFileClassLoader.java:
```diff
@@ -75,7 +87,9 @@ public JarFileClassLoader(URL[] urls, boolean verify, ClassLoader parent) {
             List<ResourceHandler> tempResourceHandlers = new ArrayList<ResourceHandler>(urls.length);
             for (URL url : urls) {
-                tempResourceHandlers.add(createResoureHandler(url, verify));
+                ResourceHandler handler = createResoureHandler(url, verify);
```
Take this opportunity to change the method name to `createResourceHandler`.

jhanders34 replied:
Done

dazavala commented on dev/com.ibm.ws.kernel.boot.core/src/com/ibm/ws/kernel/internal/classloader/DirectoryResourceHandler.java:
```diff
@@ -73,6 +76,33 @@ public Manifest getManifest() throws IOException {
+    public Set<String> getClassPackages() {
+        Set<String> packages = new HashSet<>();
+        StringBuilder sb = new StringBuilder();
+        getPackages(directory, packages, sb);
```
Once the `packages` set is computed for this class there's no need to ever compute it again on subsequent calls to method `getClassPackages()`.

jhanders34 replied:
Right.  It is only called once, so is there any concern here?  I could move the getClassPackages() function to the JarFileClassLoader instead of in the ResourceHandler methods if that would be better.

dazavala replied:
Just bringing it to your attention in case you hadn't noticed, because the method has public access.  I had considered the method is called once in this solution and didn't request a change here.  I like the current organization of the getClassPackages() behavior.  If usage changes we can always modify the method to return the previously computed set.

dazavala commented on dev/com.ibm.ws.kernel.boot.core/src/com/ibm/ws/kernel/internal/classloader/JarResourceHandler.java:
```diff
@@ -130,6 +134,35 @@ public Manifest getManifest() throws IOException {
+            return Collections.emptySet();
+        }
+
+        Enumeration<JarEntry> enumer = jarFile.entries();
```
Similar to my previous comment for `DirectoryResourceHandler.getPackages()`, once `jarFile` is set and `getClassPackages()` is called for the first time, there is no need to recompute the `packageNames` Set on subsequent calls to `getClassPackages()`.

dazavala replied:
This addressed by my previous comments to `DirectoryResourceHandler`.

dazavala commented on dev/com.ibm.ws.kernel.boot.core/src/com/ibm/ws/kernel/internal/classloader/JarFileClassLoader.java:
```diff
@@ -120,22 +140,60 @@ public void close() {
+
+        List<ResourceHandler> pkgHandlers = handlerMap.get(pkg);
+
+        if (pkgHandlers != null) {
```
And there's the crux of the solution.  I had trouble locating FAT classes that directly exercise `JarURLClassloader.addUrl()`.  Please ensure the solution works when URLs are added dynamically to either the `BootstrapChildFirstJarFileClasloader` or its base class, `JarFileClassloader`.

jhanders34 replied:
`addURL()` is called by the OSGi code.  I tried removing `addURL()` to not need to handle the dynamic portion, but it is necessary.   With the logic that I already added in `addURL()`, it does work now.  Without adding the logic to that method to call the `populateHandlerMap`, no features would start.  Only the kernel would start.

dazavala replied:
Verified to work -- thank you.

dazavala reviewed (changes requested):
Hello, @jhanders34.  I left a few comments and requested you fix method name `JarFileClassloader.createResoureHandler()`.  I didn't find any obvious errors in populating the handler map or exception handling, but it's always a good idea to review those fine details before delivery.  Trace generated for the recursion in `JarFileClassloader.getPackages()` will be interesting; I think it's valuable as it shows the building of package names, so I did not ask to `@Trivial` the method.  Finally, thanks for moving the classloader parent reference into the base class, `JarFileClassloader`.  -Regards.

dazavala reviewed (dismissed):
@jhanders34 Approved with comments.  I'll look for any subsequent pushes in case you decide to change anything.

jhanders34 committed:
Application ClassLoader optimizations

- Update loadClass to only lock around loading of the class from this
loader.  In ParentLast load don't hold the classloader lock when calling
the parent.  Also when calling the class generator don't hold the
classloader lock.  It is not needed.
- Update to use replace(char, char) instead of replaceAll(String,
String)

jhanders34 commented:
#libby #build

jhanders34 commented:
I also added @tkburroughs on review due to the change to not do locking when generating class.

dazavala commented on dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java:
```diff
@@ -693,25 +684,42 @@ protected Class<?> findOrDelegateLoadClass(String name, boolean onlySearchSelf,
+            findException = cnfe;
         }
-        return findClass(name, returnNull);
+        if (result == null) {
```
Method `generateClass(clsName)` now executes outside of the `clsName` lock.  The method defines classes w/o checking `clsName` is not already defined by first calling `findLoadedClass()`.  Previously there was no need to as any classes generated for `clsName` were defined within the lock context.  There's now risk that more than one thread may concurrently generate and define classes for the same `clsName`.  At minimum `generateClass()` should verify a generated class is not already defined before defining it; and, the call to `findLoadedClasses()` must be synchronized as it is not thread safe -- presumably you would use the lock returned by getClassloadingLock(`clsName`) for this.  These changes would still alleviate some contention on the lock for `clsName`, which I presume is the reason for the PR.

tkburroughs commented:
@jhanders34 Why is the locking not required when generating classes?  Most EJB artifacts will be done serially, so that would not be a concern, however _Stub classes for remote EJB are generated on first use; so there could be multiple threads attempting to define the same _Stub class concurrently.  Does the defineClass path already handle that?

dazavala reviewed (changes requested):
Hello, @jhanders34.  Regarding your second commit: I requested a change to `generateClass()`; beyond that I found no difference/regression in loadClass() behavior of either app classloader type.

tkburroughs reviewed (dismissed):
Latest changes that move generateClass inside a sync/lock look good to me.  Thanks

dazavala reviewed (dismissed):
Recent push 5b810943667ea91183ec06c9edd4c09a1802e316 resolves my previous review concerns.  Thanks!

dazavala reviewed (approved):
The latest commit e8be1ac9f45963186e80cf4fe50c5da6a52c8c97 LGTM.
